### PR TITLE
Inline limit in SQL sent from `dbt show`

### LINKED ISSUE #8417: [CT-2994] [Bug] `sql_header` with jinja not parsed correctly with `dbt show`
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

If you have some jinja in your `sql_header`, `dbt show` injects it verbatim, without doing any parsing - this results in invalid sql characters.

### Expected Behavior

`dbt show` should be well behaved and `sql_header` should be parsed/resolved prior to sending the query to the datawarehouse.

### Steps To Reproduce

1. Project setup:

```yml
# dbt_project.yml
name: my_dbt_project
profile: snowflake
config-version: 2
version: 1.0

# prettier-ignore
models:
  my_dbt_project:
    +materialized: table
    +sql_header: "{{ 'CALL SYSTEM$WAIT(' ~ var('wait_seconds', 0) ~ ');' }}"
```

```sql
-- models/foo.sql
select 1 id
```

2. With both dbt-core 1.5.latest (`1.5.4`) and 1.6.latest (`1.6.0`), do a show on foo:

```sh
$ dbt --debug show -s foo
23:43:46  Concurrency: 1 threads (target='default')
23:43:46  
23:43:46  Began running node model.my_dbt_project.foo
23:43:46  Re-using an available connection from the pool (formerly list_development_dbt_jyeo, now model.my_dbt_project.foo)
23:43:46  Began compiling node model.my_dbt_project.foo
23:43:46  Writing injected SQL for node "model.my_dbt_project.foo"
23:43:46  Timing info for model.my_dbt_project.foo (compile): 11:43:46.804486 => 11:43:46.817100
23:43:46  Began executing node model.my_dbt_project.foo
23:43:46  Using snowflake connection "model.my_dbt_project.foo"
23:43:46  On model.my_dbt_project.foo: /* {"app": "dbt", "dbt_version": "1.6.0", "profile_name": "snowflake", "target_name": "default", "node_id": "model.my_dbt_project.foo"} */
{{ 'CALL SYSTEM$WAIT(' ~ var('wait_seconds', 0) ~ ');' }}select 1 id
23:43:46  Opening a new connection, currently in state closed
23:43:48  Snowflake adapter: Snowflake query id: 01ae558f-0403-767c-000d-37831bf6791e
23:43:48  Snowflake adapter: Snowflake error: 001003 (42000): SQL compilation error:
syntax error line 1 at position 1 unexpected '{'.
23:43:48  Timing info for model.my_dbt_project.foo (execute): 11:43:46.820584 => 11:43:48.530728
23:43:48  On model.my_dbt_project.foo: Close
23:43:49  Database Error in model foo (models/foo.sql)
  001003 (42000): SQL compilation error:
  syntax error line 1 at position 1 unexpected '{'.
23:43:49  Finished running node model.my_dbt_project.foo
23:43:49  Connection 'master' was properly closed.
23:43:49  Connection 'model.my_dbt_project.foo' was properly closed.
23:43:49  Encountered an error:
Runtime Error
  Database Error in model foo (models/foo.sql)
    001003 (42000): SQL compilation error:
    syntax error line 1 at position 1 unexpected '{'.
23:43:49  Command `dbt show` failed at 11:43:49.175370 after 8.42 seconds
23:43:49  Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'end', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10e13b820>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x110836130>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10fe41850>]}
23:43:49  Flushing usage events
```

### Relevant log output

_No response_

### Environment

```markdown
- OS: macOS
- Python: 3.9.13
- dbt: 1.5.4 / 1.6.0
```

### Which database adapter are you using with dbt?

snowflake

### Additional Context

If we downgrade to `dbt-core==1.5.0`, show works as expected:

```sh
$ dbt --debug show -s foo
23:45:51  Concurrency: 1 threads (target='default')
23:45:51  
23:45:52  Began running node model.my_dbt_project.foo
23:45:52  Re-using an available connection from the pool (formerly list_development_dbt_jyeo, now model.my_dbt_project.foo)
23:45:52  Began compiling node model.my_dbt_project.foo
23:45:52  Writing injected SQL for node "model.my_dbt_project.foo"
23:45:52  Timing info for model.my_dbt_project.foo (compile): 11:45:52.017324 => 11:45:52.140537
23:45:52  Began executing node model.my_dbt_project.foo
23:45:52  Using snowflake connection "model.my_dbt_project.foo"
23:45:52  On model.my_dbt_project.foo: /* {"app": "dbt", "dbt_version": "1.5.0", "profile_name": "snowflake", "target_name": "default", "node_id": "model.my_dbt_project.foo"} */
select 1 id
23:45:52  Opening a new connection, currently in state closed
23:45:53  SQL status: SUCCESS 1 in 2.0 seconds
23:45:53  Timing info for model.my_dbt_project.foo (execute): 11:45:52.146470 => 11:45:53.773370
23:45:53  On model.my_dbt_project.foo: Close
23:45:54  Finished running node model.my_dbt_project.foo
23:45:54  Connection 'master' was properly closed.
23:45:54  Connection 'model.my_dbt_project.foo' was properly closed.
23:45:54  Command end result
23:45:54  Previewing node 'foo':
| ID |
| -- |
|  1 |

23:45:54  Command `dbt show` succeeded at 11:45:54.583669 after 9.60 seconds
23:45:54  Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'end', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x105ba2ce0>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x105e98190>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x119358d00>]}
23:45:54  Flushing usage events
```

But that's simply because we don't inject `sql_header` as we are now doing with #7568

----

Additionally, in the dbt Cloud IDE, 1.5.latest works because it does `show --inline` with the file contents:

image

### LINKED ISSUE #8496: [CT-3032] `dbt show --limit` should include the limit in the DWH query, to avoid long-running queries
### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

Most adapters (check impact matrix below) load the entire model into memory while running `dbt show`.
The goal of this ticket is to avoid this behavior and only load the relevant rows from the model.

### Acceptance criteria

1. Inline `limit` into the query instead of using the `fetchmany` method for `dbt show`.
2. Confirm that the query sent to Snowflake does have the limit appended to it.
3. Add a test case where there is a `limit` already in the query.

Make any needed changes to the adapter repos

### Impact to Other Teams

Adapter impact matrix

| Adapter    | Impacted? | Notes |
|------------|-----------|-------|
| Postgres   | ✅| | 
| Snowflake  | ✅| |
| BigQuery   | 🚫| |      
| Databricks | ❔| |
| Redshift   | ❔| |

### Will backports be required?

Backport to `1.6.latest`, `1.5.latest`

### Context

Original bug report in Slack: https://dbt-labs.slack.com/archives/C05FWBP9X1U/p1692909690843649.

Link to similar issue that was fixed in Core: https://github.com/dbt-labs/dbt-core/issues/7390.

The root cause of the issue is that warehouse providers are implementing the [`fetchmany`](https://peps.python.org/pep-0249/#fetchmany) method in the [DB-API 2.0](https://peps.python.org/pep-0249/#fetchmany) spec inefficiently by loading the entire table into memory and filtering client-side.

The long-term fix here is for them to become smarter in how they implement `fetchmany`, similar to what BigQuery has done: https://github.com/googleapis/google-cloud-python/issues/9185.

### PR DESCRIPTION
**Pull request by MichelleArk**
resolves #8496 
resolves #8417

### Solution

Adapter zone tests:
* dbt-bigquery: https://github.com/dbt-labs/dbt-bigquery/pull/927
* dbt-snowflake: https://github.com/dbt-labs/dbt-snowflake/pull/779
* dbt-redshift: https://github.com/dbt-labs/dbt-redshift/pull/611
* dbt-spark: TODO

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: core/dbt/include/global_project/macros/adapters/show.sql [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{% macro show(limit) -%}
```
Comment (jtcohen6) [inline core/dbt/include/global_project/macros/adapters/show.sql:1]: What do you think of giving this macro a longer name, like `get_show_sql`? (There is always some slight risk of adding a macro to the global namespace that collides with a macro which users have already defined in their projects.)

For readability, I'd suggest either:
- an inline comment that `compiled_code` is coming from [the context](https://github.com/dbt-labs/dbt-core/blob/d597b80486ebb409685fafb23d972f5361ddf7fc/core/dbt/context/providers.py#L1397)
- adding `compiled_code` and `sql_header` as macro arguments, and passing them in explicitly from `adapter.execute_macro`
```python
            kwargs={
                "compiled_code": model_context["compiled_code"],
                "sql_header": model_context["config"].get("sql_header"),
                "limit": limit
            },
        )
```

**Review by jtcohen6 — COMMENTED** [review]
Really happy about this change! It hews closer to our existing patterns, and so it's easier to reason about, and would be easier to extend with more things (beyond `limit` + `sql_header`)

I left two quick comments about the `show` macro - feel free to take 'em or leave 'em

File: core/dbt/include/global_project/macros/adapters/show.sql (hunk 1 above)
Comment (MichelleArk) [reply core/dbt/include/global_project/macros/adapters/show.sql:1]: Great suggestions, will incorporate them + go for passing in macro arguments explicitly :)

File: core/dbt/include/global_project/macros/adapters/show.sql [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{% macro get_show_sql(compiled_code, sql_header, limit) -%}
+  {%- set sql_header = sql_header -%}
+  {{ sql_header if sql_header is not none }}
+  {%- if limit is not none -%}
+  {{ get_limit_subquery_sql(compiled_code, limit) }}
```
[the remark this answers is not in the document]
Comment (MichelleArk) [reply core/dbt/include/global_project/macros/adapters/show.sql:5]: It returns all rows! There is a test for that case: https://github.com/dbt-labs/dbt-core/pull/8641/files#diff-6cd0c2e8c5c04d1c5e93c3f8916b21d8b9f626d060b4858d62de0f88ea16158aR31

the show task passes limit as `None` to this macro for any negative value of the limit cli arg: https://github.com/dbt-labs/dbt-core/pull/8641/files#diff-a210be47df8fb8ce0478d0ab960cb5901d7b550c9cd02d0b13222bbd258a5b0cR25

File: core/dbt/include/global_project/macros/adapters/show.sql [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+{% macro get_limit_subquery_sql(sql, limit) %}
+  {{ adapter.dispatch('get_limit_subquery_sql', 'dbt')(sql, limit) }}
+{% endmacro %}
+
+{% macro default__get_limit_subquery_sql(sql, limit) %}
+    select *
+    from (
+        {{ sql }}
+    ) as model_limit_subq
```
[the remark this answers is not in the document]
Comment (MichelleArk) [reply core/dbt/include/global_project/macros/adapters/show.sql:19]: yep! here's what happens without: 

> Runtime Error
  Database Error in sql_operation inline_query (from remote system.sql)
    subquery in FROM must have an alias
    LINE 5:     from (
                     ^
    HINT:  For example, FROM (SELECT ...) [AS] foo.

File: core/dbt/task/show.py [hunk 4]
```diff
@@ -23,14 +24,21 @@ def execute(self, compiled_node, manifest):
...
+            manifest=manifest,
+            context_override=model_context,
+            kwargs={
+                "compiled_code": model_context["compiled_code"],
+                "sql_header": model_context["config"].get("sql_header"),
+                "limit": limit,
+            },
+        )
         adapter_response, execute_result = self.adapter.execute(
-            compiled_node.compiled_code, fetch=True, limit=limit
```
Comment (MichelleArk) [inline core/dbt/task/show.py:32]: no longer passing limit through self.adapter.execute or making use of fetchmany, so closing https://github.com/dbt-labs/dbt-core/issues/8471 as part of this PR as well

**Review by dataders — COMMENTED** [review]
my spidey-sense tells me there's a potential UI footgun here, but the only thing I can think of is the one @colin-rogers-dbt surfaced (paraphrased):
> the addition LIMIT at the outermost level is easiest, but not all query planners are created equal in that some will not propogate the LIMIT inwards. The result being that the query will still take as long to execute as without the limit.

Another thought was that not all databases support limit, e.g. TSQL and it's `TOP` function. That said, `default__get_limit_subquery_sql()`, seems easily overridable to support the different syntax. something like the below

```sql
{% macro sqlserver__get_limit_subquery_sql(sql, limit) %}
    select top {{ limit }} *
    from (
        {{ sql }}
    ) as model_limit_subq
{% endmacro %}
```

**Pull request merged**